### PR TITLE
Deck of cards

### PR DESCRIPTION
**Pull request by jfcarter**
Deck of cards for playing any card game. Please test with two people as I need to make sure only the person holding their card can see it.

1. Use this link to set up the card table https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/Cardz/setupCardTable.js
2. When you grab a card straight from the deck another card takes its place in 1 second. While a card is in your hand you will notice that an overlay of what the card is appears a little in front of the card and the card itself changes to a question mark card (that is so only you can see your card). Other players will only see the question mark card.
3. Place the card near the opposite hand you used to grab the card and release the card to attach it to that hand. If the attach is successful you will feel some haptic feedback. This allows you to hold cards in any way you wish. 
4. To detach cards already placed in your hand, you need to move it away from the hand it is currently placed in. If a card is not attached to anyone or being grabbed then it will be viewable by everyone.
5. Click on the gray cube on the side of the table to destroy all the cards and reshuffle deck.

Note: When you grab a card there is a second lag before the next one is generated. That is intentional.

### CODE REVIEW

File: Marketplace/Cardz/Card.js [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+    _this.releaseGrab = function (entityID, args) {
+        var hand = args[0];
+        //check if the object you are holding is a card and if it isnt already in your hand
+        var props = Entities.getEntityProperties(_this.entityID);
+        var properties = JSON.parse(props.userData);
+        var checkIfCard = properties.card;
+        var held = properties.held;
+        var deckHandlerID = properties.deckHandlerID;
+        if ((checkIfCard == true) && (held == false) && (hand == "right")) {
```
Comment (misslivirose) [inline Marketplace/Cardz/Card.js]: Please refactor this so there isn't as much duplicate code in this file. There is a lot of shared code between this function and the next.

File: Marketplace/Cardz/Card.js [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            //get position of left and right hand
+            var left = MyAvatar.getLeftPalmPosition();
+            var right = props.position;
+            //get distance between right hand and left hand
+            var dx = left.x - right.x;
+            var dy = left.y - right.y;
+            var dz = left.z - right.z;
+            var dist = getDistance(dx, dy, dz);
+            //if card is within an acceptable distance to your hand then place it
+            if (dist < .10) {
```
Comment (misslivirose) [inline Marketplace/Cardz/Card.js]: Define .10 so it is not a magic number

File: Marketplace/Cardz/Deck_Handler.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+    function showOrHideCard(message) {
+        var data = JSON.parse(message);
+        if (data[0] == true) {
+            var cardName = Entities.getEntityProperties(data[1]).name;
+            var usability = {
+               textures: '{ "HiddenCardFile": "https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/DeckOfCards/DeckOfCardsTexture/' + cardName + '.jpg"}',
+            };
+            Entities.editEntity(data[1], usability);
+            if ((("CARD_" + cards[lastCard - 1]) == cardName) && (lastCard != 0)) {
```
Comment (misslivirose) [inline Marketplace/Cardz/Deck_Handler.js]: Please refactor this to avoid duplicating this much code

File: Marketplace/Cardz/Reset_Button.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        var properties = JSON.parse(props.userData);
+        //channel name
+        channelName = "reset-channel-"+ properties.deckHandlerID;
+    };
+
+    _this.startFarTrigger = function(entityID, data) {
+        //send the message
+        Messages.sendMessage(channelName, JSON.stringify(data));
+        //let player know they've hit the sign through haptic feedback
+        Controller.triggerShortHapticPulse(.9, 2);
```
Comment (misslivirose) [inline Marketplace/Cardz/Reset_Button.js]: Please define constants so they aren't magic numbers

File: Marketplace/Cardz/setupCardTable.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+//
+//  Distributed under the Apache License, Version 7.1.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+/* globals utils,SHOPPING_ENTITIES, TEMPLATES:true */
+Script.include(Script.resolvePath("./spawnCardTable.js") + "?" + Date.now());
+Script.include(Script.resolvePath("./utils.js") + "?" + Date.now());
+TEMPLATES = CARD_ENTITIES.Entities;
+//holds temporary templates and position values
+var tem;
```
Comment (misslivirose) [inline Marketplace/Cardz/setupCardTable.js]: Variable names should not be abbreviated per styling guidelines

File: Marketplace/Cardz/Card.js [hunk 6]
```diff
@@ -78,28 +86,8 @@
         var props = Entities.getEntityProperties(_this.entityID);
```
Comment (misslivirose) [inline Marketplace/Cardz/Card.js]: Only get the properties that you need from the entity in all cases of Entities.getEntityProperties

File: Marketplace/Cardz/Card.js [hunk 7]
```diff
@@ -141,101 +109,66 @@
...
-                var cardChannel = "card-channel-".concat(deckHandlerID);
-                var data = [true, _this.entityID, MyAvatar.sessionUUID];
-                Messages.sendMessage(cardChannel, JSON.stringify(data));
-                //delete an overlay if there is one
-                try {
-                    Overlays.deleteOverlay(cardOverlay);
-                    cardOverlay = undefined;
-                } catch (err) {
-                    //e
+        //check to see if you are holding a card that is not already being held
```
Comment (misslivirose) [inline Marketplace/Cardz/Card.js]: This also seems to be repeated code between left and right hand that can be reduced

**Comment by Nex-Pro** [comment]
Will test this shortly

**Comment by Nex-Pro** [comment]
### FAIL
Other players can see the card values.
Unable to hold the cards as described. If I manage to get one card stationary the next makes it fall down.

IMO we should not be loading the card back for every card. If this was redesigned so that all the card textures were on one atlas we would only need to load one texture.

**Comment by Nex-Pro** [comment]
@jfcarter @misslivirose I checked this with three other people. All could see the card values.

PS. Python is a much better language for scripting card games :)

**Comment by misslivirose** [comment]
@jfcarter please comment when this is ready for further review and testing

**Comment by jfcarter** [comment]
Ready for a retest

**Comment by Nex-Pro** [comment]
Will continue with test later.

**Comment by Nex-Pro** [comment]
Will group test this later.

**Comment by Nex-Pro** [comment]
Use this link to set up the card table https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/Cardz/setupCardTable.js

When you grab a card straight from the deck another card takes its place in 1 second. **Success**

While a card is in your hand you will notice that an overlay of what the card is appears a little in front of the card and the card itself changes to a question mark card (that is so only you can see your card). **SUCCESS**
Other players will only see the question mark card.**SUCCESS**
Place the card near the opposite hand you used to grab the card and release the card to attach it to that hand. **SUCCESS** But this is very clunky. There is no indication that the card has attached and instead they often fall.
This allows you to hold cards in any way you wish.

To detach cards already placed in your hand, you need to move it away from the hand it is currently placed in. **SUCCESS**

If a card is not attached to anyone or being grabbed then it will be viewable by everyone. **SUCCESS**
Click on the gray cube on the side of the table to destroy all the cards and reshuffle deck. **SUCCESS**

**Although this has met the test requirements there remain a number of problems.**

If a user gets cards too quickly and drops them to the table they stop being issued ( no more cards can be taken ). This is observed by all users at the same time. This can be fixed by picking up the last card that was taken and dropping it again.

The cards jump around on the table in a very unpredictable manner ( maybe need to up the damping )

It seems that when a user attempts to take a 53rd. card the undefined card is issued then no further cards. 

The reset will not remove the extra card.

**Comment by jfcarter** [comment]
Fixed all listed issues. Ready for retest.

**Comment by jfcarter** [comment]
Working on a couple changes, card game will not function for about 25 minutes

**Comment by jfcarter** [comment]
Ok, card game is back up. Realized everything was relative to the podium the deck was on instead of the deck itself. Bad design choice so I fixed it.

File: Marketplace/Card.js [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    var pulseStrength;
+    var rightHand;
+    var leftHand;
+
+    _this.preload = function(entityID) {
+        print("Loading Card script"); 
+        _this.entityID = entityID; 
+        rightHandJoint = MyAvatar.getJointIndex("RightHandMiddle1");
+        leftHandJoint = MyAvatar.getJointIndex("LeftHandMiddle1");
+        print(rightHandJoint + " " + leftHandJoint);
```
Comment (misslivirose) [inline Marketplace/Card.js]: Remove print statements

File: Marketplace/Card.js [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+        closeEnough = .1;
+        pulseStrength = .9;
+        rightHand = 0;
+        leftHand = 1;
+        var props = Entities.getEntityProperties(entityID);
```
Comment (misslivirose) [inline Marketplace/Card.js]: Don't abbreviate names -- also, naming userdata stored data "properties" is confusing

File: Marketplace/Card.js [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    _this.releaseGrab = function (entityID, args) {
+        var hand = args[0];
+        //check if the object you are holding is a card and if it isnt already in your hand
+        var props = Entities.getEntityProperties(_this.entityID);
+        var properties = JSON.parse(props.userData);
+        var checkIfCard = properties.card;
+        var held = properties.held;
+        var deckHandlerID = properties.deckHandlerID;
+        //check to see if you are holding a card that is not already being held
+        if ((checkIfCard === true) && (held === false)) {
```
Comment (misslivirose) [inline Marketplace/Card.js]: Consolidate these into a single function that you can call for each hand -- too much duplicated code

File: Marketplace/Card.js [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+                } else if (dist > closeEnough) {
+                    notCloseEnough();
+                }
+            }
+        //if you were already holding a card then unparent. If someone tries to grab the card out of your hand then do nothing
+        } else if ((checkIfCard === true) && (held === true) && (MyAvatar.sessionUUID == properties.me)) {
+            //this allows you to rearrange your cards without dropping them
+            var cardDistanceFromHand;
+            var handCardIsIn;
+            hand == "right" ? cardDistanceFromHand = MyAvatar.getLeftPalmPosition() : cardDistanceFromHand = MyAvatar.getRightPalmPosition();
```
Comment (misslivirose) [inline Marketplace/Card.js]: Line lengths should be less than 128 characters

File: Marketplace/Cardz/Card.js [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        print("Loading Card script"); 
+        _this.entityID = entityID; 
+        rightHandJoint = MyAvatar.getJointIndex("RightHandMiddle1");
+        leftHandJoint = MyAvatar.getJointIndex("LeftHandMiddle1");
+        print(rightHandJoint + " " + leftHandJoint);
+        closeEnough = .1;
+        pulseStrength = .9;
+        rightHand = 0;
+        leftHand = 1;
+        var props = Entities.getEntityProperties(entityID);
```
Comment (misslivirose) [inline Marketplace/Cardz/Card.js]: Don't abbreviate variable names, remove print statements--- why are there two versions of this script?

File: Marketplace/Cardz/Reset_Button.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    var pulseStrength;
+    var bothHands;
+
+    _this.preload = function(entityID) {
+        print("Loading reset script");
+        pulseStrength = .9;
+        bothHands = 2;
+        _this.entityID = entityID;
+        //get the id of the reset so you can find the unique channel name
+        var props = Entities.getEntityProperties(entityID);
```
Comment (misslivirose) [inline Marketplace/Cardz/Reset_Button.js]: please only request the properties you need

**Comment by jfcarter** [comment]
Ready to retest

**Comment by Nex-Pro** [comment]
Testing with an assistant.

**Comment by Nex-Pro** [comment]
### PASS this meets the test requirements.

Deck of cards for playing any card game. Please test with two people as I need to make sure only the person holding their card can see it.

Use this link to set up the card table https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/Cardz/setupCardTable.js Done.

When you grab a card straight from the deck another card takes its place in 1 second. **SUCCESS**

While a card is in your hand you will notice that an overlay of what the card is appears a little in front of the card and the card itself changes to a question mark card (that is so only you can see your card). **SUCCESS**
Other players will only see the question mark card. **SUCCESS**

Place the card near the opposite hand you used to grab the card and release the card to attach it to that hand. **Works after some practice.**

If the attach is successful you will feel some haptic feedback. **SUCCESS**
This allows you to hold cards in any way you wish. 

To detach cards already placed in your hand, you need to move it away from the hand it is currently placed in. **SUCCESS** Would suggest adding a little downward impetus as cards often simply sit in the air.

If a card is not attached to anyone or being grabbed then it will be viewable by everyone. **SUCCESS**

Click on the gray cube on the side of the table to destroy all the cards and reshuffle deck. **SUCCESS**

Note: When you grab a card there is a second lag before the next one is generated. That is intentional. **NOTED**

### FEEDBACK:

Game is hampered by distributed physics.  Cards do not settle quickly if there are more than a few. Observer sees a lot of flickering with the cards appearing in two places at once.
Placing cards in other hand is fiddly and unsatisfying. I often hit my vive controllers together.
Occasionally the card would attach then move to my shoulder. This was observed  by my magnificent assistant. I don't think I could have achieved more if I had scripted it.  So kudos to you @jfcarter

**Comment by misslivirose** [comment]
@jfcarter -- I'm still seeing some styling feedback waiting, let us know when that's ready so we can merge in!

**Comment by jfcarter** [comment]
The VR/AR lab at my school has been closed for a couple of weeks because a Hololens went missing and winter break. Don't have access to a Vive/Oculus to test anything in the meantime. It will open again on the 2nd when school starts back up. Will update then.

**Comment by jfcarter** [comment]
Everything is ready to go

**Comment by jfcarter** [comment]
Everything should be go good to go,

File: Marketplace/Cardz/Deck_Handler.js [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+                        "collidesWith": "static,dynamic",
+                        "visible": true
+                    };
+                    Entities.editEntity(temporaryIDHolder, allowPickup);
+                }, 1000);
+            }
+        }
+    }
+
+    function Shuffle(array) {
```
Comment (thoys) [inline Marketplace/Cardz/Deck_Handler.js]: final little thing, shuffle should start with a lower-case character

File: Marketplace/Cardz/Deck_Handler.js [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+        var currentIndex = array.length, temporaryValue, randomIndex;
+        // While there remain elements to shuffle...
+        while (0 !== currentIndex) {
+        // Pick a remaining element...
```
Comment (thoys) [inline Marketplace/Cardz/Deck_Handler.js]: indentation of comment should equal next line

**Pull request merged**